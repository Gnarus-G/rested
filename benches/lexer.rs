use criterion::{criterion_group, criterion_main, Criterion};
use rested::lexer;

const SOURCE: &str = r#"
set BASE_URL env("base")

[env("token")]

let string = "death"

post /todos {
   header "Authorization" env("token")
   body {
    userId: 1,
    id: 999,
    "title": `delectus aut ${string}`,
    "completed": env("base")
  }
}
"#;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("tokenize", |b| {
        b.iter(|| {
            let lexer = lexer::Lexer::new(SOURCE);
            let _: Vec<_> = lexer.into_iter().collect();
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
