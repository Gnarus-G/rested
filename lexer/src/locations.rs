use serde::Serialize;

use crate::Token;

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl Location {
    pub fn to_end_of(self, span: Span) -> Span {
        Span {
            start: self,
            end: span.end,
        }
    }

    pub fn is_before(self, location: Location) -> bool {
        if self.line == location.line {
            return self.col <= location.col;
        }
        return self.line < location.line;
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Span {
    pub start: Location,
    pub end: Location,
}

impl Span {
    pub fn new(start: Location, end: Location) -> Self {
        Self { start, end }
    }
    pub fn extend_to(self, end: Location) -> Span {
        Span {
            start: self.start,
            end,
        }
    }

    pub fn to_end_of(self, other_span: Span) -> Span {
        Span {
            start: self.start,
            end: other_span.end,
        }
    }

    pub fn width(&self) -> usize {
        let left = self.start.col;
        let right = self.end.col;

        if right > left {
            return right - left;
        }

        return left - right;
    }
}

impl<'source> Into<Span> for Token<'source> {
    fn into(self) -> Span {
        self.span()
    }
}

impl<'source> Into<Span> for &Token<'source> {
    fn into(self) -> Span {
        self.span()
    }
}

pub trait GetSpan {
    fn span(&self) -> Span;
}

impl<'source> GetSpan for Token<'source> {
    fn span(&self) -> Span {
        Span {
            start: self.start.clone(),
            end: Location {
                line: self.start.line,
                col: self.start.col + self.text.len(),
            },
        }
    }
}

pub trait GetSpanOption {
    fn get_span(&self) -> Option<Span>;
}

impl<T: GetSpan> GetSpanOption for Vec<T> {
    fn get_span(&self) -> Option<Span> {
        match (self.first(), self.last()) {
            (None, None) => None,
            (None, Some(e)) => Some(e.span()),
            (Some(e), None) => Some(e.span()),
            (Some(f), Some(l)) => Some(f.span().to_end_of(l.span())),
        }
    }
}
