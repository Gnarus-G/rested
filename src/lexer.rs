use std::ops::Range;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TokenKind {
    // keywords
    Get,
    Post,
    Header,
    Body,

    Ident,

    // literals
    StringLiteral,
    MultiLineStringLiteral,
    Url,

    // special characters
    LParen,
    RParen,
    LBracket,
    RBracket,
    End,

    //edge cases
    UnfinishedStringLiteral,
    UnfinishedMultiLineStringLiteral,
    IllegalToken,
}

#[derive(PartialEq, Clone, Copy)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Debug for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.line, self.col)
    }
}

#[derive(PartialEq)]
pub struct Token<'t> {
    pub kind: TokenKind,
    pub text: &'t str,
    pub location: Location,
}

impl<'i> std::fmt::Debug for Token<'i> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}({:?}) at {:?}", self.kind, self.text, self.location)
    }
}

trait CharaterTest {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool;
    fn is(&self, ch: u8) -> bool {
        self.passes(|&c| ch == c)
    }
}

impl CharaterTest for Option<&u8> {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool {
        match self {
            Some(c) => predicate(c),
            None => false,
        }
    }
}

#[derive(Debug)]
pub struct Lexer<'i> {
    input: &'i [u8],
    position: usize,
    cursor: Location,
}

impl<'i> Lexer<'i> {
    pub fn new(input: &'i str) -> Self {
        Self {
            input: input.as_bytes(),
            position: 0,
            cursor: Location { line: 0, col: 0 },
        }
    }

    pub fn input(&self) -> &'i str {
        std::str::from_utf8(&self.input).expect("input should only contain utf-8 characters")
    }

    fn input_slice(&self, range: Range<usize>) -> &'i str {
        std::str::from_utf8(&self.input[range]).expect("input should only contain utf-8 characters")
    }

    fn char_at(&self, position: usize) -> Option<&u8> {
        if position < self.input.len() {
            return Some(&self.input[position]);
        }
        return None;
    }

    fn ch(&self) -> Option<&u8> {
        self.char_at(self.position)
    }

    fn step(&mut self) {
        if let Some(b'\n') = self.ch() {
            self.cursor.line += 1;
            self.cursor.col = 0;
        } else {
            self.cursor.col += 1;
        };

        self.position += 1;
    }

    fn peek_char(&self) -> Option<&u8> {
        self.char_at(self.position + 1)
    }

    /// Assumes that the character at the current position, immediately before calling
    /// this function is also true for the predicate function given.
    fn read_while<P: Fn(&u8) -> bool>(&mut self, predicate: P) -> (usize, usize) {
        let start_pos = self.position;

        while self.peek_char().passes(&predicate) {
            self.step();
        }

        return (start_pos, self.position + 1);
    }

    fn skip_whitespace(&mut self) {
        while self.ch().passes(|c| c.is_ascii_whitespace()) {
            self.step();
        }
    }

    pub fn next(&mut self) -> Token<'i> {
        use TokenKind::*;

        self.skip_whitespace();

        let ch = match self.ch() {
            Some(ch) => ch,
            None => {
                return Token {
                    kind: TokenKind::End,
                    location: self.cursor,
                    text: "",
                }
            }
        };

        let t = match ch {
            b'"' if self.peek_char().is(b'"') => self.empty_string_literal(),
            b'"' => self.string_literal(),
            b'`' if self.peek_char().is(b'`') => self.empty_string_literal(),
            b'`' => self.multiline_string_literal(),
            b'(' => Token {
                kind: LParen,
                location: self.cursor,
                text: "(",
            },
            b')' => Token {
                kind: RParen,
                location: self.cursor,
                text: ")",
            },
            b'{' => Token {
                kind: LBracket,
                location: self.cursor,
                text: "{",
            },
            b'}' => Token {
                kind: RBracket,
                location: self.cursor,
                text: "}",
            },
            c if c.is_ascii_alphabetic() => self.keyword_or_identifier(),
            _ => Token {
                kind: IllegalToken,
                text: std::str::from_utf8(&self.input[self.position..self.position + 1]).unwrap(),
                location: self.cursor,
            },
        };

        self.step();

        t
    }

    fn multiline_string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;

        self.step(); //eat the opening quote

        let (s, e) = self.read_while(|&c| c != b'`');
        let string = self.input_slice(s..e);

        self.step(); //eat the closing quote

        if let None = self.ch() {
            return Token {
                kind: TokenKind::UnfinishedMultiLineStringLiteral,
                location,
                text: string,
            };
        }

        Token {
            kind: TokenKind::MultiLineStringLiteral,
            location,
            text: string,
        }
    }

    fn string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;

        self.step(); //eat the opening quote

        let (s, e) = self.read_while(|&c| c != b'"' && c != b'\n');
        let string = self.input_slice(s..e);

        self.step(); //eat the closing quote or newline character

        match self.ch() {
            Some(b'\n') | None => {
                return Token {
                    kind: TokenKind::UnfinishedStringLiteral,
                    location,
                    text: string,
                };
            }
            _ => {}
        }

        Token {
            kind: TokenKind::StringLiteral,
            location,
            text: string,
        }
    }

    fn empty_string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;
        self.step();
        Token {
            kind: TokenKind::StringLiteral,
            location,
            text: "",
        }
    }

    fn keyword_or_identifier(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|c| c.is_ascii_alphabetic());
        let string = self.input_slice(s..e);

        use TokenKind::*;

        match string {
            "get" => Token {
                kind: Get,
                location,
                text: string,
            },
            "post" => Token {
                kind: Post,
                location,
                text: string,
            },
            "header" => Token {
                kind: Header,
                location,
                text: string,
            },
            "body" => Token {
                kind: Body,
                location,
                text: string,
            },
            "http" | "https" => {
                let (.., e) = self.read_while(|&c| !c.is_ascii_whitespace());
                let s = self.input_slice(s..e);
                Token {
                    kind: Url,
                    location,
                    text: s,
                }
            }
            s => Token {
                kind: Ident,
                location,
                text: s,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use TokenKind::*;

    impl Into<Location> for (usize, usize) {
        fn into(self) -> Location {
            Location {
                line: self.0,
                col: self.1,
            }
        }
    }

    macro_rules! assert_lexes {
        ($input:literal, $tokens:expr) => {
            let mut lexer = Lexer::new($input);

            for token in $tokens {
                assert_eq!(lexer.next(), token);
            }
        };
    }

    #[test]
    fn lex_string_literals() {
        assert_lexes!(
            r#""hello""#,
            [Token {
                kind: StringLiteral,
                text: "hello",
                location: (0, 0).into()
            },]
        );

        assert_lexes!(
            r#""hello"#,
            [Token {
                kind: UnfinishedStringLiteral,
                text: "hello",
                location: (0, 0).into()
            },]
        );

        assert_lexes!(
            r#"
"hello
"world
"#,
            [
                Token {
                    kind: UnfinishedStringLiteral,
                    text: "hello",
                    location: (1, 0).into()
                },
                Token {
                    kind: UnfinishedStringLiteral,
                    text: "world",
                    location: (2, 0).into()
                }
            ]
        );

        assert_lexes!(
            r#" "" "" ``"#,
            [
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 1).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 4).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 7).into()
                }
            ]
        );

        assert_lexes!(
            r#" { "Bearer token" } "#,
            [
                Token {
                    kind: LBracket,
                    text: "{",
                    location: (0, 1).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "Bearer token",
                    location: (0, 3).into()
                },
                Token {
                    kind: RBracket,
                    text: "}",
                    location: (0, 18).into()
                }
            ]
        );

        assert_lexes!(
            r#"`
{
    stuff
}`

`
stuff"#,
            [
                Token {
                    kind: MultiLineStringLiteral,
                    text: "\n{\n    stuff\n}",
                    location: (0, 0).into()
                },
                Token {
                    kind: UnfinishedMultiLineStringLiteral,
                    text: "\nstuff",
                    location: (5, 0).into()
                }
            ]
        );
    }

    #[test]
    fn lex_get_url() {
        assert_lexes!(
            "get http://localhost",
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (0, 4).into(),
                }
            ]
        );
    }

    #[test]
    fn lex_get_url_with_header() {
        assert_lexes!(
            "get http://localhost { header \"Authorization\" \"Bearer token\" }",
            vec![
                Token {
                    kind: Get,
                    location: (0, 0).into(),
                    text: "get"
                },
                Token {
                    kind: Url,
                    location: (0, 4).into(),
                    text: "http://localhost"
                },
                Token {
                    kind: LBracket,
                    location: (0, 21).into(),
                    text: "{"
                },
                Token {
                    kind: Header,
                    location: (0, 23).into(),
                    text: "header"
                },
                Token {
                    kind: StringLiteral,
                    location: (0, 30).into(),
                    text: "Authorization"
                },
                Token {
                    kind: StringLiteral,
                    location: (0, 46).into(),
                    text: "Bearer token"
                },
                Token {
                    kind: RBracket,
                    location: (0, 61).into(),
                    text: "}"
                },
            ]
        );
    }

    #[test]
    fn lex_get_url_over_many_lines() {
        assert_lexes!(
            "get\nhttp://localhost",
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (1, 0).into(),
                }
            ]
        );

        assert_lexes!(
            r#"get 
    http://localhost 
{
}"#,
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (1, 4).into(),
                },
                Token {
                    kind: LBracket,
                    text: "{",
                    location: (2, 0).into(),
                },
                Token {
                    kind: RBracket,
                    text: "}",
                    location: (3, 0).into(),
                }
            ]
        );
    }

    #[test]
    fn lex_get_url_with_header_and_body() {
        assert_lexes!(
            r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#,
            vec![
                Token {
                    kind: Post,
                    location: (1, 0).into(),
                    text: "post"
                },
                Token {
                    kind: Url,
                    location: (1, 5).into(),
                    text: "http://localhost"
                },
                Token {
                    kind: LBracket,
                    location: (1, 22).into(),
                    text: "{"
                },
                Token {
                    kind: Header,
                    location: (2, 4).into(),
                    text: "header"
                },
                Token {
                    kind: StringLiteral,
                    location: (2, 11).into(),
                    text: "Authorization"
                },
                Token {
                    kind: StringLiteral,
                    location: (2, 27).into(),
                    text: "Bearer token"
                },
                Token {
                    kind: Body,
                    location: (3, 4).into(),
                    text: "body"
                },
                Token {
                    kind: StringLiteral,
                    location: (3, 9).into(),
                    text: "{neet: 1337}"
                },
                Token {
                    kind: RBracket,
                    location: (4, 0).into(),
                    text: "}"
                },
            ]
        );
    }

    #[test]
    fn lex_call_expression() {
        assert_lexes!(
            r#"env() env("stuff")"#,
            [
                Token {
                    kind: Ident,
                    text: "env",
                    location: (0, 0).into()
                },
                Token {
                    kind: LParen,
                    text: "(",
                    location: (0, 3).into()
                },
                Token {
                    kind: RParen,
                    text: ")",
                    location: (0, 4).into()
                },
                Token {
                    kind: Ident,
                    text: "env",
                    location: (0, 6).into()
                },
                Token {
                    kind: LParen,
                    text: "(",
                    location: (0, 9).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "stuff",
                    location: (0, 10).into()
                },
                Token {
                    kind: RParen,
                    text: ")",
                    location: (0, 17).into()
                }
            ]
        );
    }
}
