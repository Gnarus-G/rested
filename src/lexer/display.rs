use super::{locations::Location, TokenKind};

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenKind::*;
        let str = match self {
            Get => "get",
            Post => "post",
            Put => "put",
            Patch => "patch",
            Delete => "delete",
            Header => "header",
            Body => "body",
            Set => "set",
            Let => "let",
            Ident => "identifier",
            Boolean => "boolean",
            Number => "number",
            StringLiteral => "string",
            MultiLineStringLiteral => "string",
            Pathname => "pathname",
            Url => "url",
            Linecomment => "comment",
            Shebang => "#!...",
            Assign => "=",
            DollarSignLBracket => "${",
            LParen => "(",
            RParen => ")",
            LBracket => "{",
            RBracket => "}",
            LSquare => "[",
            RSquare => "]",
            Colon => ":",
            AttributePrefix => "@",
            Comma => ",",
            End => "Eof",
            UnfinishedStringLiteral => "\"...",
            UnfinishedMultiLineStringLiteral => "`...",
            IllegalToken => "illegal",
            Null => "null",
        };

        f.write_str(str)
    }
}

impl std::fmt::Display for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[{}:{}]", self.line + 1, self.col + 1)
    }
}
