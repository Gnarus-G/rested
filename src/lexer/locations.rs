use serde::Serialize;

use super::Token;

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Default)]
pub struct Position {
    pub value: usize,
    pub line: usize,
    pub col: usize,
}

impl From<Position> for Location {
    fn from(value: Position) -> Self {
        Self {
            line: value.line,
            col: value.col,
        }
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Default)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl Position {
    pub fn to_end_of(self, span: Span) -> Span {
        Span {
            start: self,
            end: span.end,
        }
    }
}

impl Location {
    pub fn is_before(self, location: Location) -> bool {
        if self.line == location.line {
            return self.col <= location.col;
        }
        self.line < location.line
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Span {
    pub start: Position,
    pub end: Position,
}

impl Span {
    pub fn new(start: Position, end: Position) -> Self {
        Self { start, end }
    }

    pub fn to_end_of(self, other_span: Span) -> Span {
        Span {
            start: self.start,
            end: other_span.end,
        }
    }

    pub fn width(&self) -> usize {
        let left = self.start.col;
        let right = self.end.col;

        if right > left {
            return right - left;
        }

        left - right
    }
}

pub trait GetSpan {
    fn span(&self) -> Span;
}

impl<'source> GetSpan for Token<'source> {
    fn span(&self) -> Span {
        Span {
            start: self.start,
            end: self.end_position(),
        }
    }
}
