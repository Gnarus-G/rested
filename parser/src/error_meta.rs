use colored::Colorize;

use lexer::Location;
use std::fmt::Display;

#[derive(Debug)]
pub struct ErrorSourceContext {
    above: Option<String>,
    line: String,
    below: Option<String>,
}

impl ErrorSourceContext {
    pub fn new(location: &Location, code: &str) -> Self {
        let line_of_token = location.line;
        let line_before = line_of_token.checked_sub(1);
        let line_after = line_of_token + 1;

        let get_line = |lnum: usize| code.lines().nth(lnum).map(|s| s.to_string());

        ErrorSourceContext {
            above: line_before.map(|lnum| get_line(lnum).expect("code is not empty")),
            line: get_line(line_of_token).expect("code is not empty"),
            below: get_line(line_after),
        }
    }
}

#[derive(Debug)]
pub struct Error<EK: Display + std::error::Error> {
    pub inner_error: EK,
    pub location: Location,
    pub message: Option<String>,
    pub context: ErrorSourceContext,
}

impl<EK: Display + std::error::Error> Error<EK> {
    pub fn new(inner_error: EK, location: Location, source_code: &str) -> Self {
        Self {
            inner_error,
            location,
            message: None,
            context: ErrorSourceContext::new(&location, source_code),
        }
    }

    pub fn with_message(mut self, msg: &str) -> Self {
        self.message = Some(msg.to_owned());
        self
    }
}

impl<Ek: Display + std::error::Error> std::error::Error for Error<Ek> {}

impl<EK: Display + std::error::Error> std::fmt::Display for Error<EK> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let formatted_error = &self.inner_error.to_string().red();
        let location = self.location.to_string().bold();

        let c = &self.context;

        if let Some(line) = &c.above {
            writeln!(f, "{line}")?
        }

        writeln!(f, "{}", c.line.bold())?;

        let indent_to_error_location = " ".repeat(self.location.col);

        let result = match &self.message {
            Some(m) => writeln!(
                f,
                "{}\u{21B3} {} {}\n{}   {}",
                indent_to_error_location,
                location,
                formatted_error,
                " ".repeat(self.location.col + location.len()),
                m.bright_red()
            ),
            None => writeln!(
                f,
                "{}\u{21B3} at {} {}",
                indent_to_error_location, location, formatted_error
            ),
        };

        if let Some(line) = &c.below {
            writeln!(f, "{line}")?;
        };

        result
    }
}
