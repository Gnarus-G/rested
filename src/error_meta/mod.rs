use crate::lexer::locations::{Location, Span};
use std::{fmt::Display, ops::Deref};

use serde::Serialize;

pub trait ErrorDisplay<D: Display + Deref<Target = str>> {
    fn formatted_error(&self) -> D;
    fn location(&self) -> D;
    fn line(&self) -> D;
    fn line_above(&self) -> Option<D>;
    fn line_below(&self) -> Option<D>;
    fn error_start(&self) -> Location;
    fn squiggle(&self) -> D;
    fn message(&self) -> Option<D>;

    fn format(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let formatted_error = self.formatted_error();
        let location = self.location();

        if let Some(line) = &self.line_above() {
            writeln!(f, "{line}")?
        }

        writeln!(f, "{}", self.line())?;
        let error_start = self.error_start();
        let indent_to_error_location = " ".repeat(error_start.col);

        writeln!(
            f,
            "{}{}\n{}\u{21B3} {} {}",
            indent_to_error_location,
            self.squiggle(),
            indent_to_error_location,
            location,
            formatted_error
        )?;

        if let Some(m) = &self.message() {
            writeln!(
                f,
                "{}   {}",
                " ".repeat(error_start.col + location.len()),
                m
            )?
        };

        if let Some(line) = self.line_below() {
            writeln!(f, "{line}")?;
        };

        Ok(())
    }
}

#[derive(Clone, PartialEq, Serialize)]
pub struct ErrorSourceContext {
    above: Option<String>,
    pub line: String,
    below: Option<String>,
}

impl ErrorSourceContext {
    pub fn new(location: &Location, code: &str) -> Self {
        let line_of_token = location.line;
        let line_before = line_of_token.checked_sub(1);
        let line_after = line_of_token + 1;

        let get_line = |lnum: usize| code.lines().nth(lnum).map(|s| s.to_string());

        ErrorSourceContext {
            above: line_before.map(|lnum| get_line(lnum).expect("code should not be empty")),
            line: get_line(line_of_token).expect("code should not be empty"),
            below: get_line(line_after),
        }
    }
}

#[derive(Clone, PartialEq, Serialize)]
pub struct ContextualError<EK: Display + std::error::Error + Clone> {
    pub inner_error: EK,
    pub span: Span,
    pub bad_token_at: Span,
    pub message: Option<String>,
    pub context: ErrorSourceContext,
}

impl<E: Display + std::error::Error + Clone> std::fmt::Debug for ContextualError<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self}")
    }
}

impl<E: Display + std::error::Error + Clone> ContextualError<E> {
    pub fn new(inner_error: E, span: Span, source_code: &str, bad_token_at: Span) -> Self {
        Self {
            inner_error,
            message: None,
            context: ErrorSourceContext::new(&span.end.into(), source_code),
            span,
            bad_token_at,
        }
    }

    pub fn with_message(mut self, msg: &str) -> Self {
        self.message = Some(msg.to_owned());
        self
    }
}

impl<E: Display + std::error::Error + Clone> ErrorDisplay<String> for ContextualError<E> {
    fn formatted_error(&self) -> String {
        self.inner_error.to_string()
    }

    fn location(&self) -> String {
        Location::from(self.span.start).to_string()
    }

    fn line(&self) -> String {
        self.context.line.clone()
    }

    fn squiggle(&self) -> String {
        "\u{2248}".repeat(self.span.width())
    }

    fn message(&self) -> Option<String> {
        self.message.clone()
    }

    fn line_above(&self) -> Option<String> {
        self.context.above.clone()
    }

    fn line_below(&self) -> Option<String> {
        self.context.below.clone()
    }

    fn error_start(&self) -> Location {
        self.span.start.into()
    }
}

impl<E: Display + std::error::Error + Clone> std::error::Error for ContextualError<E> {}

impl<E: Display + std::error::Error + Clone> std::fmt::Display for ContextualError<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.format(f)
    }
}
