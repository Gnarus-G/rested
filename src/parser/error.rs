use crate::lexer;
use crate::lexer::locations::Position;
use crate::lexer::{locations::GetSpan, Token, TokenKind};
use crate::utils::Array;

use crate::error_meta::ContextualError;

use super::{Parser, Result, TokenCheck};

impl<'source> std::fmt::Display for lexer::Token<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenKind::*;
        match self.kind {
            Url | Linecomment | IllegalToken => {
                write!(f, "{}<{}>", self.kind, self.text)
            }
            kind => write!(f, "{kind}"),
        }
    }
}

#[derive(Debug)]
pub struct Expectations<'i> {
    source_code: &'i str,
    pub start: Position,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum ParseError<'source> {
    ExpectedToken {
        found: lexer::Token<'source>,
        expected: TokenKind,
    },
    ExpectedEitherOfTokens {
        found: lexer::Token<'source>,
        expected: Array<TokenKind>,
    },
}

impl<'source> std::error::Error for ParseError<'source> {}

impl<'source> std::fmt::Display for ParseError<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let formatted_error = match self {
            ParseError::ExpectedToken {
                expected, found, ..
            } => {
                format!("expected '{}' but got {}", expected, found)
            }
            ParseError::ExpectedEitherOfTokens {
                found, expected, ..
            } => {
                let expected = expected
                    .iter()
                    .map(|kind| format!("'{}'", kind))
                    .collect::<Vec<String>>()
                    .join(",");
                format!("expected either one of {} but got {}", expected, found)
            }
        };

        f.write_str(&formatted_error)
    }
}

#[derive(Debug)]
pub struct ParserErrors<'source> {
    pub errors: Vec<ContextualError<ParseError<'source>>>,
}

impl<'source> ParserErrors<'source> {
    pub fn new(errors: Vec<ContextualError<ParseError<'source>>>) -> Self {
        Self { errors }
    }
}

impl<'source> std::error::Error for ParserErrors<'source> {}

impl<'source> std::fmt::Display for ParserErrors<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for err in self.errors.iter() {
            write!(f, "{err}")?
        }
        Ok(())
    }
}

impl<'i> Expectations<'i> {
    pub fn new(parser: &Parser<'i>) -> Self {
        Self {
            source_code: parser.lexer.input(),
            start: parser.curr_token().start,
        }
    }

    pub fn expect_peek<'p>(
        &self,
        parser: &'p mut Parser<'i>,
        expected_kind: TokenKind,
    ) -> Result<'i, &'p Token<'i>> {
        if parser.peek_token().is(expected_kind) {
            return Ok(parser.next_token());
        }

        let error = self.expected_token(parser.next_token(), expected_kind);

        Err(error.into())
    }

    pub fn expect_peek_one_of(
        &self,
        parser: &mut Parser<'i>,
        expected_kinds: &[TokenKind],
    ) -> Result<'i, ()> {
        if parser.peek_token().is_one_of(expected_kinds) {
            return Ok(());
        }

        let error = self.expected_one_of_tokens(parser.next_token(), expected_kinds);

        Err(error.into())
    }

    pub fn expected_token(
        &self,
        token: &Token<'i>,
        expected: TokenKind,
    ) -> ContextualError<ParseError<'i>> {
        ContextualError::new(
            ParseError::ExpectedToken {
                found: token.clone(),
                expected,
            },
            self.start.to_end_of(token.span()),
            self.source_code,
        )
    }

    pub fn expected_one_of_tokens(
        &self,
        token: &Token<'i>,
        expected_kinds: &[TokenKind],
    ) -> ContextualError<ParseError<'i>> {
        let mut expected_dedpuded: Vec<TokenKind> = vec![];

        for kind in expected_kinds {
            if !expected_dedpuded.contains(kind) {
                expected_dedpuded.push(*kind)
            }
        }

        ContextualError::new(
            ParseError::ExpectedEitherOfTokens {
                found: token.clone(),
                expected: expected_dedpuded.into(),
            },
            self.start.to_end_of(token.span()),
            self.source_code,
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::Parser;

    use insta::assert_ron_snapshot;

    macro_rules! assert_ast {
        ($input:literal) => {
            let mut parser = Parser::new($input);
            let ast = parser.parse();

            insta::with_settings!({
                 description => $input
            }, {
                assert_ron_snapshot!(ast)
            });

            assert!(!ast.errors().is_empty())
        };
    }

    #[test]
    fn expected_url_after_method() {
        assert_ast!("get {}");

        assert_ast!("post");
    }

    #[test]
    fn expected_name_after_header_keyword() {
        assert_ast!("post http://localhost {header}");
    }

    #[test]
    fn expecting_identifier_or_string_lit_after_header_name() {
        assert_ast!(r#"get http://localhost { header "name" }"#);
    }

    #[test]
    fn expecting_request_or_other_attribute_after_attributes() {
        assert_ast!(
            r#"
            @skip
            @dbg
            let k = "v"
            get http://localhost { header "name" k }"#
        );
    }

    #[test]
    fn expecting_commas_between_certain_json_items() {
        assert_ast!(
            r#"let o = {
                 yo: "joe"
                 hello: "world"
               }"#
        );
        assert_ast!(r#" let o = ["joe" "world"] "#);
    }

    #[test]
    fn expecting_partial_block_with_error() {
        assert_ast!(r#"get /hello { header "test" "value" header }"#);
    }

    #[test]
    fn expecting_partial_block_with_missing_body_expr() {
        assert_ast!(
            r#"
get /sdf {
   header "" s
   body  }
"#
        );
    }

    #[test]
    fn expecting_partial_block_with_errors() {
        assert_ast!(
            r#"
get /adsf {
  header
  body a
}
"#
        );
    }
}
