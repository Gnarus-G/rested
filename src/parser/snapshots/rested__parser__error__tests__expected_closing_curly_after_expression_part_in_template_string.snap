---
source: src/parser/error.rs
description: "`wowee ${\"hello\"} error ahead ${env(\"base\")`"
expression: ast
---
Program(
  source: "`wowee ${\"hello\"} error ahead ${env(\"base\")`",
  items: [
    Expr(TemplateStringLiteral(
      span: Span(
        start: Position(
          value: 0,
          line: 0,
          col: 0,
        ),
        end: Position(
          value: 44,
          line: 0,
          col: 44,
        ),
      ),
      parts: [
        StringPart(StringLiteral(
          raw: "wowee ",
          value: "wowee ",
          span: Span(
            start: Position(
              value: 1,
              line: 0,
              col: 1,
            ),
            end: Position(
              value: 6,
              line: 0,
              col: 6,
            ),
          ),
        )),
        ExpressionPart(String(StringLiteral(
          raw: "\"hello\"",
          value: "hello",
          span: Span(
            start: Position(
              value: 9,
              line: 0,
              col: 9,
            ),
            end: Position(
              value: 15,
              line: 0,
              col: 15,
            ),
          ),
        ))),
        StringPart(StringLiteral(
          raw: " error ahead ",
          value: " error ahead ",
          span: Span(
            start: Position(
              value: 17,
              line: 0,
              col: 17,
            ),
            end: Position(
              value: 29,
              line: 0,
              col: 29,
            ),
          ),
        )),
        ExpressionPart(Call(CallExpr(
          identifier: Ok(Token(
            kind: Ident,
            text: "env",
            start: Position(
              value: 32,
              line: 0,
              col: 32,
            ),
          )),
          arguments: ExpressionList(
            span: Span(
              start: Position(
                value: 35,
                line: 0,
                col: 35,
              ),
              end: Position(
                value: 42,
                line: 0,
                col: 42,
              ),
            ),
            items: [
              This(String(StringLiteral(
                raw: "\"base\"",
                value: "base",
                span: Span(
                  start: Position(
                    value: 36,
                    line: 0,
                    col: 36,
                  ),
                  end: Position(
                    value: 41,
                    line: 0,
                    col: 41,
                  ),
                ),
              ))),
            ],
          ),
        ))),
        ExpressionPart(Error(ContextualError(
          inner_error: ExpectedToken(
            found: Token(
              kind: OpeningBackTick,
              text: "`",
              start: Position(
                value: 43,
                line: 0,
                col: 43,
              ),
            ),
            expected: RBracket,
          ),
          span: Span(
            start: Position(
              value: 0,
              line: 0,
              col: 0,
            ),
            end: Position(
              value: 43,
              line: 0,
              col: 43,
            ),
          ),
          message: None,
          context: ErrorSourceContext(
            above: None,
            line: "`wowee ${\"hello\"} error ahead ${env(\"base\")`",
            below: None,
          ),
        ))),
      ],
    )),
  ],
)
